momepy/functional/_diversity.py

```diff
@@ -104,16 +165,112 @@ def _describe(values, q, include_mode=False):
         graph._adjacency.index.codes[0]
     )
 
-    if q is None:
-        stat_ = grouper.agg(["mean", "median", "std", "min", "max", "sum"])
-        if include_mode:
-            stat_["mode"] = grouper.agg(lambda x: _mode(x.values))
+    return _compute_stats(grouper, q, include_mode)
+
+
+def describe_reached(
+    y, graph_index, result_index=None, graph=None, q=None, include_mode=False
+) -> DataFrame:
+    """
+    Calculates statistics of ``y`` objects reached on a neighbourhood graph.
+    Requires a ``graph_index`` that links the ``y`` objects to ``graph`` or streets
+    assigned beforehand (e.g. using :py:func:`momepy.get_nearest_street`).
+    The number of elements within neighbourhood defined in ``graph``. If
+    ``graph`` is ``None``, it will assume topological distance ``0`` (element itself)
+    and ``result_index`` is required in order to arrange the results.
+    If ``graph``, the results are arranged according to the spatial weights ordering.
+    The statistics calculated are count, sum, mean, median, std.
+    Optionally, mode can be calculated, or the statistics can be calculated in
+    quantiles ``q``.
+
+    Adapted from :cite:`hermosilla2012` and :cite:`feliciotti2018`.
+
+    Parameters
+    ----------
+    y : DataFrame | Series | numpy.array
+        A GeoDataFrame containing objects to analyse.
+    graph_index : Series | numpy.array
+        The unique ID that specifies the aggregation
+        of ``y`` objects to ``graph`` groups.
+    result_index : pd.Index (default None)
+        An index that specifies how to order the results.
+    graph : libpysal.graph.Graph (default None)
+        A spatial weights matrix of the streets.
+    q : tuple[float, float] | None, optional
+        Tuple of percentages for the percentiles to compute. Values must be between 0
+        and 100 inclusive. When set, values below and above the percentiles will be
+        discarded before computation of the average. The percentiles are computed for
+        each neighborhood. By default None.
+    include_mode : False
+        Compute mode along with other statistics. Default is False. Mode is
+        computationally expensive and not useful for continous variables.
+
+    Returns
+    -------
+    DataFrame
+    """
+
+    if Version(pd.__version__) <= Version("2.1.0"):
+        raise NotImplementedError("Please update to a newer version of pandas.")
+
+    if not HAS_NUMBA:
+        warnings.warn(
+            "The numba package is used extensively in this function to accelerate the"
+            " computation of statistics but it is not installed or  cannot be imported."
+            " Without numba, these computations may become slow on large data.",
+            UserWarning,
+            stacklevel=2,
+        )
+
+    if (result_index is None) and (graph is None):
+        raise ValueError(
+            "One of result_index or graph has to be specified, but not both."
```

Comment:
Let's declare this `ValueError` so it can be used here and below. Or at least declare the message to re-use.